### ISSUE TITLE
max number of clients reached ?

### ISSUE THREAD

**Issue by onekiloparsec** [hishnash/djangochannelsrestframework#13]
Hi, 

I am actively using your project in mine, and I would like to help improving it. So far, I am not really able to contribute, as I don't understand all the things. I just wanted to report you a problem I have, and suggest a (crude) solution.

My Django app is running on Heroku with the minimum of resources (free when possible). I am using the free plan of RedisCloud with 30 connections, but still, it seems I am reaching the max.

```
Traceback (most recent call last): 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/celery/app/trace.py", line 382, in trace_task 
Mar 06 22:41:40    R = retval = fun(*args, **kwargs) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/celery/app/trace.py", line 641, in __protected_call__ 
Mar 06 22:41:40    return self.run(*args, **kwargs) 
Mar 06 22:41:40  File "/app/arcsecond/activities/tasks.py", line 52, in parse_ESO_archive_latest_rows 
Mar 06 22:41:40    target_name=row.target_name) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/django/db/models/manager.py", line 82, in manager_method 
Mar 06 22:41:40    return getattr(self.get_queryset(), name)(*args, **kwargs) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/django/db/models/query.py", line 413, in create 
Mar 06 22:41:40    obj.save(force_insert=True, using=self.db) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/django/db/models/base.py", line 718, in save 
Mar 06 22:41:40    force_update=force_update, update_fields=update_fields) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/django/db/models/base.py", line 758, in save_base 
Mar 06 22:41:40    update_fields=update_fields, raw=raw, using=using, 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/django/dispatch/dispatcher.py", line 175, in send 
Mar 06 22:41:40    for receiver in self._live_receivers(sender) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/django/dispatch/dispatcher.py", line 175, in <listcomp> 
Mar 06 22:41:40    for receiver in self._live_receivers(sender) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/djangochannelsrestframework/observer/observer.py", line 154, in post_save_receiver 
Mar 06 22:41:40    **kwargs 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/djangochannelsrestframework/observer/observer.py", line 221, in post_change_receiver 
Mar 06 22:41:40    **kwargs 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/djangochannelsrestframework/observer/observer.py", line 233, in send_messages 
Mar 06 22:41:40    async_to_sync(channel_layer.group_send)(group_name, message) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/asgiref/sync.py", line 64, in __call__ 
Mar 06 22:41:40    return call_result.result() 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/concurrent/futures/_base.py", line 425, in result 
Mar 06 22:41:40    return self.__get_result() 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result 
Mar 06 22:41:40    raise self._exception 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/asgiref/sync.py", line 78, in main_wrap 
Mar 06 22:41:40    result = await self.awaitable(*args, **kwargs) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/channels_redis/core.py", line 559, in group_send 
Mar 06 22:41:40    async with self.connection(self.consistent_hash(group)) as connection: 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/channels_redis/core.py", line 742, in __aenter__ 
Mar 06 22:41:40    self.conn = await self.pool.pop() 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/channels_redis/core.py", line 49, in pop 
Mar 06 22:41:40    conns.append(await aioredis.create_redis(**self.host, loop=loop)) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/aioredis/commands/__init__.py", line 178, in create_redis 
Mar 06 22:41:40    loop=loop) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/aioredis/connection.py", line 129, in create_connection 
Mar 06 22:41:40    await conn.auth(password) 
Mar 06 22:41:40  File "/app/.heroku/python/lib/python3.7/site-packages/aioredis/util.py", line 48, in wait_ok 
Mar 06 22:41:40    res = await fut 
Mar 06 22:41:40aioredis.errors.MaxClientsError: ERR max number of clients reached 
```

When going right into the problematic line (line 233 of `observer/observer.py`), I see a loop over `group_names`. My very crude proposition would be to extract `async_to_sync(channel_layer.group_send)` and reuse it for all `group_names`. Would it save connections ? 

Thanks again.

**Comment by onekiloparsec** [comment]
You'll be happy that by applying the proposed modification, the MaxNumberClient error does not show up anymore. Not sure I actualy fixed something, but at least, I can go on with that...

**Comment by hishnash** [comment]
Interesting, this seems like something that should also be an issue with Channels (the framework this is build on).

I will try to take a peak into what might be happening here.

**Comment by hishnash** [comment]
Looks like this was more of a redis config issue so closing, feel free to re-open if this appears again.

**Comment by onekiloparsec** [comment]
Hi. The problem's still here... I've increased the number of background tasks, and since then, i hit again the same problem. I agree this is certainly related to Django's Redis, as other people seem to have similar issues: https://stackoverflow.com/questions/52201570/django-channels-redis-channel-layer-opens-a-lot-of-connections

But my gut's feeling says it may also be amplified by djangochannelsrestframework ? I'l try to apply again the workaround mentionned above (and that I removed some time after this discussion).

**Comment by onekiloparsec** [comment]
More information here: https://github.com/django/channels_redis/issues/125

Which triggers that question: do we really need `async_to_sync` in that place in the `observer.py` ?

**Comment by onekiloparsec** [comment]
Sorry for the message rafale... It seems that the very latest version of channel_redis `2.3.1` contains an import cleanup connections fix. See here. https://github.com/django/channels_redis/pull/130

I'll monitor the thing for a while, but that's good news. In the meantime, I'll leave my question above open to your comments. Cheers.

**Comment by hishnash** [comment]
Django's `Signal` system calls in as sync methods. The `handle` method that sends the msg cant therefore be async. 

We could look into created a singleton async run-loop for the observers and have the `handle` method put the msg into a queue that the run loop consumes from and then injects into Redis. 

Let me know if this issue arises again. If it does possibly I can create a general sync `send` method that used the above-mentioned singleton async run loop to send the message.

**Comment by onekiloparsec** [comment]
I applied my workaround and the django_redis fix at the same time, and the result is stunning. Connections number oscillates between 2 and 3! Instead of rising indefinitely up to 256... 

I just removed my workaround (which nonetheless kept the sync aspect of the code, it only avoided the loop of over the groups), and we'll see what happens. I'll keep you posted, but I suspect the django_redis fix was the key.
